### DISCUSSION TITLE
How to more easily diagnose the reason for Bad Request

### DISCUSSION THREAD

**Discussion by sparrowt** [carltongibson/django-filter#1676] [Q&A] [answered]
Today I spent a long time trying to figure out why a caller was getting a 400 response as the only thing in the logs was this:
```
Bad Request: /my/api/path/
```
from Django's `log_response`. This was on a remote server so I couldn't just attach a debugger.

I eventually realised it was due to the filter being passed e.g. `/my/api/path/foo=bar` where `bar` was not one of the value choices for foo, but this was only by trial and error.

This is the basics of the code in question:
```python
class MyFilterSet(FilterSet):
    class Meta:
        model = MyModel
        fields = ["foo"]

    foo = MultipleChoiceFilter(choices=MyFooType.choices)

class MyViewSet(viewsets.ReadOnlyModelViewSet):
    filter_backends = [DjangoFilterBackend]
    filterset_class = MyFilterSet
    ...
```

Whilst typing this I've realised that the actual response _body_ probably explains exactly what the problem is and so the problem really is that the _client_ (in my case another cloud service) should be logging not just "I got a 400 response" but also the response body at that point.

So this may just be a "PEBKAC" but I'll post it anyway in case it helps someone else.

Just in case though, is there any way to get the _server_ to log more detail as to _why_ it is about to return a 400?

If there were a generic way to log the reason for any validation error on the server that would really help - though perhaps this is a DRF question rather than django-filter specifically...

(I guess one could write middleware which catches `HttpResponseBadRequest` and then logs & re-raises? but that feels overkill)

**Comment by carltongibson** [comment, marked as answer]
Hi @sparrowt. 

You could subclass the ['django_filters.rest_framework.DjangoFilterBackend'](https://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend) and override `filter_queryset()` to catch the `rest_framework.exceptions.ValidationError` and log the details. 

That would be pretty light-weight (and well targeted) I'd suggest. 

Have fun ⛱️

**Reply by sparrowt** [reply]
Thank you, really appreciate the reply and that makes sense 👍I realise it's not in the scope of this repo but if I can find something more generic to do the same for all DRF validation errors (not just django-filter specific) I think that would be even more useful

**Reply by carltongibson** [reply]
You'd have to intercept DRF's exception handling, and add logging in the right places.

**Reply by sparrowt** [reply]
Hm yeah there's [EXCEPTION_HANDLER](https://www.django-rest-framework.org/api-guide/exceptions/#custom-exception-handling) but those DRF docs explicitly says that it won't handle some of the validation cases that I'm interested in:
> Note that the exception handler will only be called for responses generated by raised exceptions. It will not be used for any responses returned directly by the view, such as the HTTP_400_BAD_REQUEST responses that are returned by the generic views when serializer validation fails.

however I'm struggling to find the code for those 'generic views' it talks about.

There's also the ability to override Django [`handler400`](https://docs.djangoproject.com/en/dev/topics/http/views/#customizing-error-views) but again I'm not convinced this is called in the DRF ValidationError case.

I'll need to dig more!
